torch/distributed/rpc/rref_proxy.py

```diff
@@ -12,27 +14,60 @@ def _local_invoke(rref, func_name, args, kwargs):
 def _local_invoke_async_execution(rref, func_name, args, kwargs):
     return getattr(rref.local_value(), func_name)(*args, **kwargs)
 
+# convert a Future[Future[X]] to a Future[X]
+def _unwrap_future(fut: Future) -> Future:
+    res: Future = Future()
+
+    def _unwrap_level0(fut0):
+        try:
+            res.set_result(fut0.wait())
+        except BaseException as ex:
+            res.set_exception(ex)
+
+    def _unwrap_level1(fut1):
+        try:
+            fut1.wait().then(_unwrap_level0)
+        except BaseException as ex:
+            res.set_exception(ex)
+
+    fut.then(_unwrap_level1)
+    return res
+
 def _invoke_rpc(rref, rpc_api, func_name, timeout, *args, **kwargs):
-    # Since rref._get_type can potentially issue an RPC, it should respect the
-    # passed in timeout here.
-    rref_type = rref._get_type(timeout=timeout)
-
-    _invoke_func = _local_invoke
-    # Bypass ScriptModules when checking for async function attribute.
-    bypass_type = issubclass(rref_type, torch.jit.ScriptModule) or issubclass(
-        rref_type, torch._C.ScriptModule
-    )
-    if not bypass_type:
-        func = getattr(rref_type, func_name)
-        if hasattr(func, "_wrapped_async_rpc_function"):
-            _invoke_func = _local_invoke_async_execution
-
-    return rpc_api(
-        rref.owner(),
-        _invoke_func,
-        args=(rref, func_name, args, kwargs),
-        timeout=timeout
-    )
+    # this function returns either the value or a Future depending on what rpc_api is.
+    def _rref_type_cont(future, is_future):
+        if is_future:
+            rref_type = future.wait()
+        else:
+            rref_type = future
+
+        _invoke_func = _local_invoke
+        # Bypass ScriptModules when checking for async function attribute.
+        bypass_type = issubclass(rref_type, torch.jit.ScriptModule) or issubclass(
+            rref_type, torch._C.ScriptModule
+        )
+        if not bypass_type:
+            func = getattr(rref_type, func_name)
+            if hasattr(func, "_wrapped_async_rpc_function"):
+                _invoke_func = _local_invoke_async_execution
+
+        return rpc_api(
+            rref.owner(),
+            _invoke_func,
+            args=(rref, func_name, args, kwargs),
+            timeout=timeout
+        )
+
+    request_blocking = rpc_api != rpc_async
+    rref_val = rref._get_type(timeout=timeout, blocking=request_blocking)
+
+    # _get_type return type depends on how the first call is made to it
+    # so it's possible to get a Future even if you call it with `blocking=True`
+    is_future = issubclass(type(rref_val), torch._C.Future)
+    if request_blocking or not is_future:
+        return _rref_type_cont(rref_val, is_future=is_future)
+    else:
+        return _unwrap_future(rref_val.then(lambda fut: _rref_type_cont(fut, is_future=True)))
```

Comment:
Shouldn't `return rref_val.then(..)` be sufficient here? Why do we need the unwrapping?

Reply:
`then` produces a Future to the return value of its callback and he callback in this case is rpc_async is produces a Future itself, so we'd end up with a `Future[Future[T]]` to the target.

Reply:
Ah I see, so instead of doing this unwrapping can we do this instead:

1. Create a future `res` in this function and pass it into the function `_rref_type_cont`. 
2. Then inside `_rref_type_cont`, assign the return value of `rpc_api` to `res`.
3. Finally return `res` from this method.

Reply:
Done, I inlined _unwrap_future in its call site.